enclave/core/sgx/exception.c

```diff
@@ -219,8 +219,61 @@ int _emulate_illegal_instruction(sgx_ssa_gpr_t* ssa_gpr)
     // Emulate CPUID
     if (*((uint16_t*)ssa_gpr->rip) == OE_CPUID_OPCODE)
     {
-        return oe_emulate_cpuid(
+        int ret = oe_emulate_cpuid(
             &ssa_gpr->rax, &ssa_gpr->rbx, &ssa_gpr->rcx, &ssa_gpr->rdx);
+        if (ret == 0)
+            ssa_gpr->rip += 2;
+        return ret;
+    }
+
+    // Emulate wrfsbase
+    const uint32_t OE_WRFSBASE_PREFIX_1 = 0xae0f48f3;
+    if (*((uint32_t*)ssa_gpr->rip) == OE_WRFSBASE_PREFIX_1)
+    {
+        uint64_t regs[] = {
+            ssa_gpr->rax, // d0
+            ssa_gpr->rax, // d1
+            ssa_gpr->rdx, // d2
+            ssa_gpr->rbx, // d3
+            ssa_gpr->rsp, // d4
+            ssa_gpr->rbp, // d5
+            ssa_gpr->rsi, // d6
+            ssa_gpr->rdi, // d7
+        };
+
+        uint8_t last_byte = ((uint8_t*)ssa_gpr->rip)[4];
+        uint8_t idx = last_byte - 0xd0;
+        if (idx >= sizeof(regs) / sizeof(regs[0]))
+            return -1;
+
+        ssa_gpr->fs_base = regs[idx];
+        ssa_gpr->rip += 5;
+        return 0;
+    }
+
+    // Emulate wrfsbase
+    const uint32_t OE_WRFSBASE_PREFIX_2 = 0xae0f49f3;
+    if (*((uint32_t*)ssa_gpr->rip) == OE_WRFSBASE_PREFIX_2)
+    {
+        uint64_t regs[] = {
+            ssa_gpr->r8,  // d0
+            ssa_gpr->r9,  // d1
+            ssa_gpr->r10, // d2
+            ssa_gpr->r11, // d3
+            ssa_gpr->r12, // d4
+            ssa_gpr->r13, // d5
+            ssa_gpr->r14, // d6
+            ssa_gpr->r15, // d7
+        };
+
+        uint8_t last_byte = ((uint8_t*)ssa_gpr->rip)[4];
+        uint8_t idx = last_byte - 0xd0;
+        if (idx >= sizeof(regs) / sizeof(regs[0]))
+            return -1;
+
+        ssa_gpr->fs_base = regs[idx];
+        ssa_gpr->rip += 5;
+        return 0;
```

Comment:
Nit: this code is the same as the above code right? Can this be factored into `oe_emulate_wrfsbase()`. Also would rdfsbase also need to be emulated?

Feel free to ignore until we decide if the approach will be accepted or not.

Reply:
The registers are different. If found this to be more readable than if-else.

Reply:
We could emulate `rdfsbase` if needed. It is possible.

Reply:
Oh I see. That's fine then.